### ISSUE TITLE
Restarting sender does not update the queue length metric on the receiver

### ISSUE THREAD

**Issue by indualagarsamy** [Particular/NServiceBus.Metrics.ServiceControl#27]
Problem:

When all the endpoints are run the first time, the ServicePulse Monitoring dashboard reports queue length correctly. 

Restarting the sending and receiving endpoint, and running the same load test, the dashboard shows all of the other metrics, such as Processing time, Critical time, etc increasing, but queue length remains at 0. 

@mikeminutillo was able to repro this on the monitoring demo after updating to RC bits.

**Comment by mikeminutillo** [comment]
This is happening because the QueueLength Tracking is using a stable/deterministic sessionId. You can see this here https://github.com/Particular/NServiceBus.Metrics.ServiceControl/blob/master/src/NServiceBus.Metrics.ServiceControl/ServiceControlReporting/QueueLengthTracker.cs#L33

What happens is that the sender updates the count using this sessionId and the monitoring server records the max that has been sent and the max that has been received.

So we get:

```
(Session1, Sent 5)
(Session1, Sent10)
(Session1, Sent15)
(Session1, Received 10)
```

At this point QueueLength is 5 (15 sent - 10 received)

Now we restart the sender but it comes back with the same session id so we get

```
(Session1, Sent 5)
(Session1, Sent 7)
(Session1, Received 5)
```

Because the monitoring server is only updating to the max of either, the queue length _still says 10_. 

This situation won't get better until we surpass the original counts. At that point queue length will report properly again.

We _should_ be changing the sessionId every restart. OR we need a separate session key that changes every run. I need to check with the @Particular/metrics-maintainers why this isn't the case. I suspect it might have something to do with removing old counters in the receiving endpoint and the monitoring server.

**Comment by mikeminutillo** [comment]
Changing the session key on restart doesn't solve the problem.

The Queue Length calculation done by the monitoring server is based off of establishing a link between a sending endpoint and a receiving endpoint. It uses session key to do this. 

If you have a backlog of messages from the old session, the messages from the new session will not get counted until you receive the first one. We need to establish the links independently of the sessions. Here is what I proposed on Slack:

---

When sending a message we send (SessionKey-Destination) as the counter key. We should probably send (EndpointName-Destination) as the stable "Logical Message Source" and (SessionKey, Counter) as the volatile part.

When receiving we can establish there's a link between Logical Message Source and Logical Receiver (Receiving Endpoint Name). We keep a dictionary of counters under that link keyed off of SessionKey.

To find the queue length of an endpoint we need to compare the number of messages that have been added to the Logical Message Source (across all sessions) against the number of messages that have been processed by the Logical Receiver (again across all sessions).

Two downsides.
1) How the heck do we handle the old messages that aren't formed that way?
2) How do we clean up old session data?
